### PR TITLE
Add un-prefix docker labels for dockershim

### LINKED ISSUE #47368: Add Optional Label Metadata to Pods' Containers
It should be possible for the user to apply custom labels to a pod's containers, and have those labels propagate down into docker.  
This would be helpful for interacting with non-k8s-specific programs that access docker directly.

Related: #3764

Outdated: #13513  (I don't want to fight a reopen-war, but this probably should not have been closed.)

### PR DESCRIPTION
**Pull request by mhumeSF**
```release-note
Add un-prefix docker labels for dockershim
```

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Small addition that, when running a kubelet that uses docker runtime, adds un-prefixed annotations as container labels in addition to the prefixed annotations. This is useful for sidecar containers that look for certain container labels, but are unable to because of the additional prefix added. 

**Which issue(s) this PR fixes**:
Fixes #47368

### CODE REVIEW

**Comment by mhumeSF** [comment]
I signed it

**Review by mattjmcnaughton — APPROVED** [review]
I can't see any harm in this and it does sound like it could be useful.
/ok-to-test

/assign @vishh

**Review by mattjmcnaughton — CHANGES_REQUESTED** [review]
I think the test failure here is legit - https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/82973/pull-kubernetes-bazel-test/1176207891477041159

I think we need to update tests that check for annotations to expect this new annotation.

**Comment by mhumeSF** [comment]
I'll update

**Comment by mhumeSF** [comment]
Ended up having labels have their own prefix so that the they could be extracted from the extractlabels functions. The other option would be to update the extractLabels function to test if a label exists with both an annotationPrefix and without and adding to annotations list; but I think this is ok in it's current form. 🤷‍♂️

**Comment by alejandrox1** [comment]
@mhumeSF this looks great!
While you're working on this  you should checkout in particular the output from the `pull-kubernetes-bazel-build` job. 
You can also try building locally by doing something like `bazel build //cmd/kubelet:kubelet` 

/priority important-soon
/milestone v1.17

**Review by mattjmcnaughton — COMMENTED** [review]
Apologies if you already discussed this - can you share more about whether we need to/should update the unit tests after this change? I'm a little surprised to see them passing after your change, given that no corresponding updates were made to the unit tests.

**Comment by mhumeSF** [comment]
The test I saw failing was `TestLabelsAndAnnotationsRoundTrip` from `pkg/kubelet/dockershim/helpers_test.go`. This was due to the kubelet annotations getting turned into labels in the extractLabels function. I added a conditional to check for dupe labels without the `annotation.` prefix and ignore those when adding to label maps. Let me know if you think this is an ok solution.

**Comment by mattjmcnaughton** [comment]
> The test I saw failing was `TestLabelsAndAnnotationsRoundTrip` from `pkg/kubelet/dockershim/helpers_test.go`. This was due to the dupe kubelet annotations incorrectly getting turned into labels because they didn't have the annotation prefix in the extractLabels function. I added a conditional to check for dupe labels without the `annotation.` prefix and ignore those when adding to label maps. This works yea?

Tbh, I'm not sure I have enough context on k8s expectations for labels and annotations to answer that :) Maybe @vishh (tagged here as the approver) will know :)

/assign @vishh

**Comment by mhumeSF** [comment]
@vishh @mattjmcnaughton Any estimate on when this could be reviewed?

**Comment by mattjmcnaughton** [comment]
I'll defer to @vishh here :)

**Comment by smourapina** [comment]
@vishh @mattjmcnaughton Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on November 18. Is this PR still on track for 1.17?

**Comment by mattjmcnaughton** [comment]
@mhumeSF it might be worth asking in #sig-node on slack if there is another `OWNER` who can take a look at this :)

**Comment by smourapina** [comment]
@mattjmcnaughton @mhumeSF I just pinged #sig-node on Slack!

**Comment by derekwaynecarr** [comment]
Can we have a KEP or a design that better describes the use case?  The sig is in the process of transitioning everything through the CRI and I would prefer we not have special dockershim-specific behavior without a discussion about how its being used in containerd/cri-o/etc.

File: pkg/kubelet/dockershim/helpers.go
Comment (derekwaynecarr) [inline pkg/kubelet/dockershim/helpers.go]: The referenced issue claims:

"This is useful for sidecar containers that look for certain container labels, but are unable to because of the additional prefix added."

Can you elaborate on the use case?  What is the side car container?  Is the container deployed in the same pod?  Why does the container depend on these labels versus normal intra-pod communication?

More detail on the actual use case would help so we can properly transition use cases for users that adopt containerd/cri-o or any other runtime.

Comment (mhumeSF) [reply pkg/kubelet/dockershim/helpers.go]: This would help with services like registrator or traefik where it is looking for certain docker labels, but because of the annotation prefix sidecar services are unable to pickup docker labels added to the containers.

Comment (mhumeSF) [reply pkg/kubelet/dockershim/helpers.go]: Probably not using the term sidecar correctly. More so services looking for docker labels

Comment (derekwaynecarr) [reply pkg/kubelet/dockershim/helpers.go]: So containers deployed on host outside of Kubernetes?  Not an actual “sidecar” container which typically means a peer Container in same pod?

Comment (mhumeSF) [reply pkg/kubelet/dockershim/helpers.go]: No all containers are deployed via kubernetes manifests

Comment (mhumeSF) [reply pkg/kubelet/dockershim/helpers.go]: But yes, not a container in the same pod.

Comment (mhumeSF) [reply pkg/kubelet/dockershim/helpers.go]: So if traefik or registrator ran as a daemonset on a node but using docker labels for service discovery purposes

**Comment by mhumeSF** [comment]
https://github.com/kubernetes/kubernetes/issues/47368 also describes scenarios where this PR would be useful

https://github.com/kubernetes/kubernetes/issues/47368#issuecomment-308455270 
> This is so that external/management systems can query docker for this data, without needing to be aware of k8s.

**Comment by smourapina** [comment]
@mhumeSF @derekwaynecarr Friendly reminder: code freeze is rapidly approaching and will take place in November 14th. Is this PR still intended for 1.17? Otherwise we will need to move it to 1.18 milestone.

**Comment by smourapina** [comment]
Hello @mhumeSF and @derekwaynecarr! Bug Triage team here for the 1.18 release. This is a friendly reminder that code freeze is scheduled for 5 March. 
Will there be progress on this PR to still make it to milestone 1.18? Thanks in advance!

**Comment by mhumeSF** [comment]
@josiahbjorgaard @smourapina I'll close for now.

**Pull request closed without merging**
